aw_transform/transforms.py

```diff
@@ -0,0 +1,174 @@
+    """
+
+    events = sorted(events, key=lambda e: e.timestamp)
+    filterevents = sorted(filterevents, key=lambda e: e.timestamp)
+    filtered_events = []
+
+    e_i = 0
+    f_i = 0
+    while e_i < len(events) and f_i < len(filterevents):
+        event = events[e_i]
+        filterevent = filterevents[f_i]
+        ep = _get_event_period(event)
+        fp = _get_event_period(filterevent)
+
+        ip = ep.intersection(fp)
+        if ip:
+            # If events itersected, add event with intersected duration and try next event
+            filtered_events.append(_replace_event_period(event, ip))
+            e_i += 1
+        else:
+            # No intersection, check if event is before/after filterevent
+            if ep.end <= fp.start:
+                # Event ended before filter event started
+                e_i += 1
+            elif fp.end <= ep.start:
+                # Event started after filter event ended
+                f_i += 1
+            else:
+                logger.warning("Unclear if/how this could be reachable, skipping period")
+                e_i += 1
+                f_i += 1
+
+    return filtered_events
+
+
+def filter_keyvals(events, key, vals, exclude=False) -> List[Event]:
+    def predicate(event):
+        for val in vals:
+            if key in event.data and val == event.data[key]:
+                return True
+        return False
+
+    if exclude:
+        return list(filter(lambda e: not predicate(e), events))
+    else:
+        return list(filter(lambda e: predicate(e), events))
+
+
+def merge_events_by_keys(events, keys):
+    # The result will be a list of events without timestamp since they are merged
+    # Call recursively until all keys are consumed
+    if len(keys) < 1:
+        return events
+    merged_events = {}
+    for event in events:
+        summed_key = ""
+        for key in keys:
+            if key in event.data:
+                summed_key = summed_key + "." + event["data"][key]
+        if summed_key not in merged_events:
+            merged_events[summed_key] = deepcopy(event)
+            merged_events[summed_key].data = {}
+            for key in keys:
+                if key in event.data:
+                    merged_events[summed_key].data[key] = event.data[key]
+        else:
+            merged_events[summed_key].duration += event.duration
+    result = []
+    for key in merged_events:
+        result.append(Event(**merged_events[key]))
+    return result
+
+def sort_by_timestamp(events):
+    return sorted(events, key=lambda e: e.timestamp)
+
+def sort_by_duration(events):
+    return sorted(events, key=lambda e: e.duration, reverse=True)
+
+def limit_events(events, count):
+    return events[:count]
+
+"""
+    Watcher specific transforms
+"""
+
+def split_url_events(events):
+    for event in events:
+        if "url" in event.data:
+            url = event.data["url"]
+            protocol_end = url.find('://')
```

Comment:
You should definitely be using [`urllib.parse`](https://docs.python.org/3.6/library/urllib.parse.html?highlight=urllib%20parse#module-urllib.parse) instead.

Reply:
fixed